# Task Exercise 4.8 - Grouping and Aggregating Instacart data variables

## The Instacart team wants to derive information from the Instacart Basket dataframe pertaining to customer spending habits, and to target marketing to specific customer subsets, including customer loyalty, rates of spending, and frequent vs. infrequent customers.  

## This script contains the following points:

### 1. Import and verify 'orders_products_merged.pkl' dataframe

### 2. Task 4 - Using transform and loc() to identify loyalty customers on full dataframe and creating loyalty labels by user id corresponding to their order value ('Loyal customer', 'Regular customer', 'New customer')*

### 3. Task 2 - Grouping average number of orders per department ID using the aggregate function

### 4. Task 5 - Check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer)

### 5. Task 6 - Create a spending flag for each user based on the average price across all their orders

### 6. Task 7 - Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 

### 9. Export as 'orders_products_merged.pkl'

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

### 1. Import and verify 'orders_products_merged.pkl' dataframe

In [2]:
path = r'C:\Users\howl6\OneDrive\Certificates\CareerFoundry\Coursework\Data_Immersion\Chapter 4\Instacart Basket Analysis'

In [3]:
# Import 'orders_products_merged.pkl' dataframe

ords_prods_merge = pd.read_pickle(os.path.join(path,'02_Data','Prepared_Data', 'orders_products_merged.pkl'))

 ### 2. *NOTE - the transform and loc functions for Task 4 were run in a prior notebook and saved previously in the orders_products_merged dataframe for this exercise ('4.8 IC Grouping Data & Aggregating Variables'), resulting in the creation of the max_order and loyalty_flag columns below.

In [4]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  first_time_order  add_to_cart_order  reordered  _merge  \
0  ...             False                  5          0    both   
1  ...             False                  1          1    both   
2  ...             False                 20          0    both   
3  ...              True                 10          0    both   
4  ...             False                 11          1    both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy           Most orders   
2  Mid-range product     Busiest day    Busiest days        Average orders   
3  Mid-range product  Regularly busy    Slowest days           Most orders   
4  Mid-range product      Least busy    Slowest days           Most orders   

  max_order      loyalty_flag  
0        32  Regular customer  
1        32  Regular customer  
2         5      New customer  
3         3      New customer  
4         3      New customer  

[5 rows x 21 columns]

### 3. Task 2 - In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [5]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Task 3 - The results were similar between the full dataset and the subset of 1000000 rows, with differences primarily noted on the right side of the decimal point.  Of note, the mean for Department_ID 21 was lower in comparison to the subset (22.90 vs. 25.53) and Department_ID 10 was higher in comparison to the subset (20.2 vs. 18.68).

### 4. Task 5 - The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [6]:
# Descriptive statistics of 'prices' on 'loyalty_flag'

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

###  Based on the mean prices for each of the customer categories, and appears that the 'Loyal customer' pays the lowest average price in comparision to the 'Regular customer' and 'New customer'.

### 5. Task 6 - The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [7]:
# Identify the mean price by user id.

ords_prods_merge['mean_prices'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [8]:
pd.options.display.max_rows = None

In [9]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [10]:
# Create spending flags for 'Low spender' and 'High spender'

ords_prods_merge.loc[ords_prods_merge['mean_prices'] < 10, 'spending_flag'] = 'Low spender'

In [11]:
ords_prods_merge.loc[ords_prods_merge['mean_prices'] >= 10, 'spending_flag'] = 'High spender'

In [12]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770746
High spender      634113
Name: spending_flag, dtype: int64

### The majority of spenders appear to spend less than 10 on average.

In [13]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered  _merge    price_range_loc     busiest_day    busiest_days  \
0  ...          0    both  Mid-range product  Regularly busy  Regularly busy   
1  ...          1    both  Mid-range product  Regularly busy  Regularly busy   
2  ...          0    both  Mid-range product     Busiest day    Busiest days   
3  ...          0    both  Mid-range product  Regularly busy    Slowest days   
4  ...          1    both  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag mean_prices  spending_flag  
0           Most orders        32  Regular customer    6.935811    Low spender  
1           Most orders        32  Regular customer    6.935811    Low spender  
2        Average orders         5      New customer    7.930208    Low spender  
3           Most orders         3      New customer    4.972414    Low spender  
4           Most orders         3      New customer    4.972414    Low spender  

[5 rows x 23 columns]

### Task 7 - In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [14]:
# Identify the median order by user id.

ords_prods_merge['median_days_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [27]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [24]:
# Create frequency flags for 'non-frequent customer', and 'High spender'

ords_prods_merge.loc[ords_prods_merge['median_days_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [25]:
ords_prods_merge.loc[(ords_prods_merge['median_days_prior_order'] <= 20) & (ords_prods_merge['median_days_prior_order'] > 10), 'frequency_flag'] = 'Regular customer'

In [22]:
ords_prods_merge.loc[ords_prods_merge['median_days_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [26]:
ords_prods_merge.head(100)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

In [28]:
ords_prods_merge.to_pickle(os.path.join(path, '02_Data','Prepared_Data', 'orders_products_merged.pkl'))